Sources/Sentry/SentryProfiler.mm

```diff
@@ -242,146 +240,193 @@
-    if (_gCurrentProfiler->_spansInFlight.count == 0) {
-        SENTRY_LOG_DEBUG(@"Stopping profiler %@ because span with id %@ was last being profiled.",
-            _gCurrentProfiler, span.spanId.sentrySpanIdString);
-        [self stopProfilerForReason:SentryProfilerTruncationReasonNormal];
+    if (![_gCurrentProfiler isRunning]) {
+        SENTRY_LOG_WARN(@"Current profiler is not running.");
+        return;
     }
+
+    [self stopProfilerForReason:SentryProfilerTruncationReasonNormal];
 }
 
-+ (void)dropTransaction:(SentryTransaction *)transaction
++ (BOOL)isRunning
 {
     std::lock_guard<std::mutex> l(_gProfilerLock);
-
-    const auto spanID = transaction.trace.spanId;
-    const auto profiler = _gProfilersPerSpanID[spanID];
-    if (profiler == nil) {
-        SENTRY_LOG_DEBUG(@"No profiler tracking span with id %@", spanID.sentrySpanIdString);
-        return;
-    }
-
-    [self captureEnvelopeIfFinished:profiler spanID:spanID];
+    return [_gCurrentProfiler isRunning];
 }
 
-+ (void)linkTransaction:(SentryTransaction *)transaction
++ (SentryEnvelopeItem *)createProfilingEnvelopeItemForTransaction:(SentryTransaction *)transaction
 {
     std::lock_guard<std::mutex> l(_gProfilerLock);
 
-    const auto spanID = transaction.trace.spanId;
-    SentryProfiler *profiler = _gProfilersPerSpanID[spanID];
-    if (profiler == nil) {
-        SENTRY_LOG_DEBUG(@"No profiler tracking span with id %@", spanID.sentrySpanIdString);
-        return;
+    if (_gCurrentProfiler == nil) {
+        SENTRY_LOG_DEBUG(@"No profiler from which to receive data.");
+        return nil;
     }
 
-    SENTRY_LOG_DEBUG(@"Found profiler waiting for span with ID %@: %@",
-        transaction.trace.spanId.sentrySpanIdString, profiler);
-    [profiler addTransaction:transaction];
+    NSMutableDictionary<NSString *, id> *payload = [_gCurrentProfiler->_profileData mutableCopy];
 
-    [self captureEnvelopeIfFinished:profiler spanID:spanID];
-}
+    NSArray *samples = ((NSArray *)payload[@"profile"][@"samples"]).mutableCopy;
+    if ([samples count] < 2) {
+        SENTRY_LOG_DEBUG(@"Not enough samples in profile");
+        return nil;
+    }
 
-+ (BOOL)isRunning
-{
-    std::lock_guard<std::mutex> l(_gProfilerLock);
-    return [_gCurrentProfiler isRunning];
+    // slice the profile data to only include the samples/metrics within the transaction
+    const auto slicedSamples = [self slicedArray:samples transaction:transaction];
+    if (slicedSamples.count < 2) {
+        SENTRY_LOG_DEBUG(@"Not enough samples in profile during the transaction");
+        return nil;
+    }
+    payload[@"profile"][@"samples"] = slicedSamples;
+
+    // add the serialized info for the associated transaction
+    const auto firstSampleTimestamp
+        = (uint64_t)[slicedSamples.firstObject[@"elapsed_since_start_ns"] longLongValue];
+    const auto profileDuration = getDurationNs(firstSampleTimestamp, getAbsoluteTime());
```

Comment:
@indragiek are we good with this? I have a subsequent PR that will change everything to be relative to the transaction start instead of the profile start, which will change what we use for the start in this calculation. The options to calculate a profile duration would then be:
1. transaction start to last profiler sample
2. transaction start to when the profiler is actually stopped (which can be up to 9.9 milliseconds longer than option 1)

Reply:
I think we should just leave this as-is for this PR and address it when I tackle https://github.com/getsentry/team-profiling/issues/183

Reply:
+1 to leaving the current behavior unchanged and handling it as part of the separate change to normalize the timestamps since this PR is already big enough 

@philipphofmann I generally agree that using the last sample timestamp would be more accurate but since this is a behavior change from how it works currently I think it's out of scope for this PR.

Reply:
Thanks for the clarification, @armcknight. I'm okay with following up on this.